include/AnnotationHandler.h

```diff
@@ -0,0 +1,42 @@
+#ifndef ANNOTATION_HANDLER_H
+#define ANNOTATION_HANDLER_H
+
+#include "Annotation.h"
+#include "Debug.h"
+#include "ErrorAnnotation.h"
+#include "FunctionAnnotation.h"
+#include "ParameterAnnotation.h"
+#include "ReturnAnnotation.h"
+#include "StructAnnotation.h"
+#include "Utils.h"
+
+class AnnotationHandler {
```

Comment:
This class should have an explicit comment that indicates that it represents the annotations _in a particular file_ (based on reading the code, I think that's true). If it's not true, we're going to have problems with disambiguating names.

Reply:
When you had asked me about it in the meeting I hadn't thought about whether this was going to be for a file or for a program, but after thinking about it more, it seems best that it'll be for each file. I'll start writing tests that analyze several files in a directory

Reply:
My only concern about having an `AnnotationHandler` for each file is about `#include` directives: if a header file with annotations is included in another file that's being analyzed, will the annotations show up in the `AnnotationHandler` for the file under analysis, or will it need to be linked to the header file's `AnnotationHandler` somehow?

Reply:
I have found out that the annotations do not show up in the IR of the file that's being analyzed. This is actually quite a problem since if we call some annotated function that's from some other .h file, there are 0 hints in the IR that even indicate where the function comes from. The #include directives aren't even listed in the IR so we'd have to look at the .c file itself to find them. I have been researching to see if there's some way to include everything in the #include directives into the IR for some files to no success. I'm quite unsure how this should be handled, so what do you suggest?

Reply:
Yes, this is a very serious problem. I'm not sure what the "right" way to do this in C is.

Here is one possible solution:
* first, the analysis makes a quick (AST only) pass over all input files, only looking for annotations. It creates an internal model (using something like `AnnotationHandler`) of every annotated function in the program.
* then, we run our existing analysis.

We could also separate out the first part as a separate program that produced "annotation files" or something similar that our analysis knows how to read. I'm not sure that has any real benefits, though.

Reply:
I think with this AST pass we'll also have to see what the files include as well since there's still the issue of file disambiguation. I like the idea of generating an annotation file for our analysis to read; I'm thinking it could tell us, for some annotation, which files that annotation applies to so that we know we're applying the correct one when we are analyzing a specific file.

Reply:
"AST" refers to an [Abstract Syntax Tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree), which is the output of a parser. Some part of LLVM should be able to produce one from C code, though I'm not sure which part.